### PR TITLE
Adds support for generic location costs.

### LINKED ISSUE #9: Stag stations don't cost Geo, but should
StagStation placement handler should place vanilla cost for the Stag station and force the player to have to buy it to unlock it.

### LINKED ISSUE #21: Basin fountain should cost money
Basin fountain doesn't cost money for some people, although this is possibly inconsistent.

### LINKED ISSUE #31: Don't randomize shop costs for vanilla shop placements.
If an item is not randomized and it normally belongs in a shop, do not randomize its Geo price.

Example: Turning off RandomizeCharms should not randomize geo costs of charms in Leg Eater's shop.

This might require tracking all the items which are within a shop by default. It could also mean not placing the vanilla item and then including the DefaultShopItem in the ShopLocation to have that shop's vanilla items.

### LINKED ISSUE #73: Rework shop costs as a generic "cost overrides" mechanism
Currently, shop values all have separate handlers for logic and placement rules.

The proposal here is to include a generic "cost overrides" dictionary in SlotData that might have a format that looks something akin to this:

```
  "costOverrides": {
    "grubfather_3": [{"amount": 5, "currency": "grubs"}],
    "salubra_requires_charms_1": [{"amount": 5, "currency": "charms"}, {"amount": 450, "currency": "geo"}]
  }
```

This would include costs for stag stations, maps, the fountain check and all other checks where costs are involved and would better facilitate future YAML options that scale geo costs by a percentage.

### PR DESCRIPTION
**Pull request by dewiniaid**
AP can now include location_costs in slotdata, of the form
```{location: {currency: cost, currency: cost...}}```

If location_costs are present, previous placement handler logic is ignored
(it's only kept for backwards compatibility) and the location costs is
the sole determining factor of all costs.

This fixes vanilla costs (stags, maps, fountain), as well as enabling the
possibility of... weird cost randomization options going in the future
like stag stations requiring grubs and what not.

Oh, also, we reverse the location list before adding it, as a quick and
dirty fix to get things sorted in the right order.

Closes #73, closes #31, closes #21, closes #9

### CODE REVIEW

**Comment by dewiniaid** [comment]
Also closes #16

File: Archipelago.HollowKnight/SlotData/SlotDataExtract.cs [hunk 1]
```diff
@@ -18,5 +19,16 @@ internal static class SlotDataExtract
...
             return extractedObject;
         }
+
+        public static Dictionary<string, Dictionary<string, int>> ExtractLocationCostsFromSlotData(object v)
+        {
+            Dictionary<string, Dictionary<string, int>> ret = new();
+            var jobj = ExtractObjectFromSlotData<Dictionary<string, JObject>>(v);
+            foreach (var key in jobj.Keys)
+            {
+                ret[key] = jobj[key].ToObject<Dictionary<string, int>>();
```
Comment (Ijwu) [inline Archipelago.HollowKnight/SlotData/SlotDataExtract.cs:29]:
```suggestion
                ret[key] = ExtractObjectFromSlotData<Dictionary<string, int>>(jobj[key]);
```
`ExtractObjectFromSlotData` just has some null safety that's nice to have around, so might as well use it.

File: Archipelago.HollowKnight/ArchipelagoRandomizer.cs [hunk 2]
```diff
@@ -59,20 +62,35 @@ internal class ArchipelagoRandomizer
...
+                placementHandlers = new List<IPlacementHandler>()
+                {
+                    new ShopPlacementHandler(Random),
+                    new GrubfatherPlacementHandler(GrubfatherCosts),
+                    new SeerPlacementHandler(SeerCosts),
+                    new EggShopPlacementHandler(EggCosts),
+                    new SalubraCharmShopPlacementHandler(SalubraCharmCosts, Random)
+                };
+            }
+            //LocationCosts = SlotDataExtract.ExtractLocationCostsFromSlotData(slotData["location_costs"]);
```
Comment (Ijwu) [inline Archipelago.HollowKnight/ArchipelagoRandomizer.cs:92]: Remove comment if unnecessary.

File: Archipelago.HollowKnight/SlotData/SlotDataExtract.cs (hunk 1 above)
Comment (dewiniaid) [reply Archipelago.HollowKnight/SlotData/SlotDataExtract.cs:29]: We can do that, but it ends up logging a *ton* of garbage to the log.  Or maybe it's time to remove some of that debug logging (along the item placement logging)

File: Archipelago.HollowKnight/ArchipelagoRandomizer.cs [hunk 3]
```diff
@@ -208,23 +226,78 @@ public void PlaceItem(string location, string name, NetworkItem netItem)
...
+                    Archipelago.Instance.LogWarn($"Found zero cost types when handling placement at location {originalLocation}!");
+                    return;
+                }
+                var costTag = item.AddTag<CostTag>();
+                if (costs.Count == 1)
+                {
+                    costTag.Cost = costs[0];
+                } else
+                {
+                    costTag.Cost = new MultiCost(costs);
```
[the remark this answers is not in the document]
Comment (dewiniaid) [reply Archipelago.HollowKnight/ArchipelagoRandomizer.cs:294]: Only if it does not support costs at all (some of those might exist), in which case this probably does nothing.  Noticeably, `ISingleCostPlacement` instances won't even look at the cost tag and instead look at their own cost attribute.

I don't know what happens if you string together a long set of costs (geo + grubs + essence + etc), but MultiCost appears to support it and itself it's just an instance of Cost.

Not all placements will give you the **option** to pay a cost though -- I'm told (but haven't verified) that CostChestPlacements will just take the geo from you if you have it whether you want to spend the geo or not.

[the remark this answers is not in the document]
Comment (dewiniaid) [reply Archipelago.HollowKnight/ArchipelagoRandomizer.cs:294]: In any event, that's not something we need to address until we add fancy cost rando to AP, and it's likely something that'll be addressed on the AP side (though that might involve generating a json file of locations and what they support).

File: Archipelago.HollowKnight/SlotData/SlotDataExtract.cs (hunk 1 above)
Comment (Ijwu) [reply Archipelago.HollowKnight/SlotData/SlotDataExtract.cs:29]: Debug logging being chatty isn't a bad thing.

**Pull request merged**